### ISSUE TITLE
Hash of a github release changes because of versioneer

### ISSUE THREAD

**Issue by MarkWieczorek** [python-versioneer/python-versioneer#217]
**Problem:**

I created a github release, which makes use of versioneer to generate a version number. I downloaded the tar.gz archive and computed the hash (which is required by third-party package managers, like macports and homebrew-core). A few days later, however, I was notified that the hash of the released archive file **changed**.

Inspecting the original and subsequent tar.gz files, I found that the tar.gz archive did indeed change, and the cause of this change is `git_refnames` in the file `_version.py` generated by versioneer.

```
diff -r SHTOOLS-4.7.1 SHTOOLS2-4.7.1
diff -r SHTOOLS-4.7.1/pyshtools/_version.py SHTOOLS2-4.7.1/pyshtools/_version.py
26c26
<     git_refnames = " (HEAD -> master, tag: v4.7.1)"
---
>     git_refnames = " (HEAD -> master, tag: v4.7.1, develop)"
```
The only thing that makes sense is that Github **recreated** the project tar.gz file a few days later, and that for some reason, the git tag was modified somehow. Though I don't know why github would recreate this file, I also question whether `git_refnames` is required by versioneer. and `_version.py`.

**Additional context:**
The release of our project, along with the release tag "v4.7.1" was created using the Github release page. I then pulled the tags to my local repository using `git pull --tags`. We work on a parallel `develop` branch, and have since pushed new commits to the develop branch on Github.

**Examples of other people having the same problem:**
* https://github.com/jborg/attic/issues/335
* https://github.com/spack/spack/issues/7937
* https://trac.macports.org/ticket/41657#comment:2
* https://github.com/macports/macports-ports/pull/8518

**Solution**
I don' know the solution, nor if there is one. I have opened an issue at GiHub support to find out why the tar.gz file was recreated after the release: This seems very dangerous to me, but perhaps there is a reason this was done.

If there is not a solution, then at a minimum, versioneer should provide a warning in the documentation as to how to **avoid** this problem. It is not 100% clear what I could have done differently to avoid this.

**Comment by ryandesign** [comment]
> The only thing that makes sense is that Github **recreated** the project tar.gz file a few days later, and that for some reason, the git tag was modified somehow. Though I don't know why github would recreate this file,

GitHub generates the tarballs on-demand whenever they're requested. GitHub might cache them for a short time, I don't know, but they're definitely not kept forever. If you want to distribute files that are guaranteed never to change unless you change them, upload those files to the GitHub release.

**Comment by MarkWieczorek** [comment]
But I did download the files from the GitHub release page...
![Screen Shot 2020-09-29 at 16 31 02](https://user-images.githubusercontent.com/8697285/94572531-6e4c8280-0271-11eb-842a-52e0f714faf6.png)

**Comment by ryandesign** [comment]
Yes, but those are not release downloads. There is a distinction between automatically generated tarballs, which those are, and the manually-uploaded release files which would appear on lines after those two.

**Comment by ryandesign** [comment]
Consider https://github.com/macports/macports-base/releases/tag/v2.6.3 where you can see the many files manually uploaded to the release in addition to the two "Source code" links at the end which are automatically generated on demand.

**Comment by MarkWieczorek** [comment]
For info, I got a response from github, and they confirm that each time you download a tarball archive, that it is regenerated. Because of this, the code in versioneer

```
git_refnames = "$Format:%d$"
```
Is updated with each download which can give rise to seemingly trivial problems where branches are added or removed from the tagged commit like
```
git_refnames = " (HEAD -> master, tag: v4.7.1)"
```
```
git_refnames = " (HEAD -> master, tag: v4.7.1, develop)"
```
As far as I can tell, versioneer only cares about the "tag: vX.X" part in git_refnames, so it would be great if there was a way to write only info without the branch names. I looked quickly through the git archive documentation, but it doesn't look like there is a format specifier that does this. **Is there a different way to write only the tag name to the _versioneer.py` file?**

**Comment by effigies** [comment]
No, unfortunately this is just how `git archive` works, providing all refnames for the commit.

It looks like you're setting up a process to basically snapshot the initial archive. If you'd like to avoid the possibility of changes, you can also modify your process to never push a branch head and tag together. e.g.,

```
git tag -a vX.Y.Z
# Make some small change or commit with --allow-empty
git commit -m 'post-tag bump'
git push origin master
```

Another option is to use an uploaded sdist as your source distribution, rather than a git archive. You could fetch it from PyPI or add it to your GitHub release. Those do only have tag and commit information in them, and are not subject to the variability inherent to git archive.

**Comment by MarkWieczorek** [comment]
Thanks. I'll probably just try to use github actions to upload a tarball archive to the release. If you know of a github action that does this, let me know (I haven't been able to find one that does this out of the box).

**Comment by grembo** [comment]
@MarkWieczorek @effigies Maybe this is helpful to you (using git 2.32+): https://github.com/isamert/scli/issues/151#issuecomment-1004846067

Not sure if it helps in your case though (I don't know enough about versioneer, how it creates _version.py and which consequences changing it would have).

**Comment by effigies** [comment]
That's awesome! Any chance of duplicating your test repository on Gitlab? That's the main other public source. And we should see what the failure case looks like from a manual `git archive`, as I'm sure that's in somebody's process.

**Comment by grembo** [comment]
@effigies I tried on gitlab, seems to work as well: https://gitlab.com/grembo/gitformattest/-/tags

Failures I've seen:
- git too old: No substitution happens at all (so it's `<hash> %(describe...)`
- nothing to describe (no annotated tags): Replaced by blank (so its `<hash> `)

**Comment by effigies** [comment]
I [see you saw](https://github.com/effigies/gitformattest/commit/f34897b18214e5760fcc872bac5dc298c0582cf3) `%(descibe:tags=true)` in dev, and we'll see how long that takes to hit GitHub/GitLab. Unfortunately without lightweight tag support, I think this will end up less reliable than the current approach.

In the short term, a possibility would be to add a configuration option to switch to `%(describe)` instead of `%d`, which would avoid breaking existing setups that do use lightweight tags.

**Comment by grembo** [comment]
> I [see you saw](https://github.com/effigies/gitformattest/commit/f34897b18214e5760fcc872bac5dc298c0582cf3) `%(descibe:tags=true)` in dev, and we'll see how long that takes to hit GitHub/GitLab. Unfortunately without lightweight tag support, I think this will end up less reliable than the current approach.
> 
> In the short term, a possibility would be to add a configuration option to switch to `%(describe)` instead of `%d`, which would avoid breaking existing setups that do use lightweight tags.

Probably the best approach for the time being.

**Comment by exquo** [comment]
Side note: a convenient one-liner to get the contents of the `archive`'d `VERSION` file:

	git archive main | tar -Oxf - VERSION
